NOAA-OWP/t-route
# Data Assimilation (DA) for the diffusive wave model(s)

# Executive Summary
This PR adds the capability to do data assimilation (DA) with the diffusive wave model. At gage locations, simulated flows are either replaced with interpolated USGS observations, or nudged by some amount (in the event that observations are missing). Replaced/nudged flows at gage locations are then routed downstream. The DA scheme does not propagate any gage information upstream. Instead gage information (flow) is only passed in the downstream direction. The approach used for observation interpolation, DA substitution, and DA decay is exactly the same as is used with Muskingum-Cunge. 

# Technical Details
- DA with the diffusive wave model is facilitated by subnetwork creation. The spatial unit of analysis for the diffusive wave Fortran kernel is a stand-alone stream network, as opposed to a single segment in the case of Muskingum-Cunge. Therefore, it is necessary to break networks into subnetworks that reside between gages and reservoirs. Subnetwork tailwaters are either gages, reservoir inflows, or network tailwaters. This allows downstream flow conditions from one subnetwork to be passed as upstream boundary conditions to the subsequent subnetwork (either a stream network or reservoir). Subnetwork creation between gages and reservoirs is done by revisions in [`nhd_network.build_subnetworks_btw_reservoirs`](https://github.com/awlostowski-noaa/t-route/blob/5c9fed318caeef2588131b8c2b1da49ee8af3422/src/python_framework_v02/troute/nhd_network.py#L610-L765). When handed a list of gages in the network (from `usgs_df`) the function builds subnetworks between reservoirs and gages. Otherwise, the function only build subnetworks between reservoirs. 
- The DA substitution and decay processes are very similar to that used with Muskingum-Cunge, in that they both use the [`simple_da.simple_da_with_decay`](https://github.com/awlostowski-noaa/t-route/blob/5c9fed318caeef2588131b8c2b1da49ee8af3422/src/python_routing_v02/troute/routing/fast_reach/simple_da.pyx#L81-L117) function called from a cython compute kernel. However, the DA scheme differs from Muskingum-Cunge in that the diffusive DA occurs [AFTER](https://github.com/awlostowski-noaa/t-route/blob/5c9fed318caeef2588131b8c2b1da49ee8af3422/src/python_routing_v02/troute/routing/fast_reach/diffusive_cnt.pyx#L255-L315) all segments have been simulated for all timesteps. 
- [Changes to `diffusive_utils.py`](https://github.com/awlostowski-noaa/t-route/blob/5c9fed318caeef2588131b8c2b1da49ee8af3422/src/python_routing_v02/troute/routing/diffusive_utils.py#L500-L523) allow multiple off-network upstream inflows to a single subnetwork headwater as an upstream boundary condition. Gages frequently exist on junction inflow segments. So, the segment immediately below that junction (subnetwork headwater), needs to receive boundary condition flows from the gage segment and all other segments influent to the junction.  

# Demonstration
The figure below demonstrates diffusive wave data assimilation in action. The mainstem of the Eno River, between Hillsborough and Durham, North Carolina has two gages. Gage segments are colored red. Above the upstream most gage at Hillsborough, the open loop simulation is shown - this solution is replaced at the gage segment near Hillsborough. Here, we can see that the assimilation routine is properly replacing routed flows with interpolated observations. Between the two gages, the assimilated observations are routed downstream. At the downstream gage, the simulations are again replaced with interpolated observations.
![diffusive_da_demo_eno](https://user-images.githubusercontent.com/66840445/133653435-fb71dee5-5ad8-4aad-96f0-1795b1732d7a.png)


# Testing
1. Ensure that simulation results are stable throughout the domain - no negative or NaNs in the result arrays
- Results are stable. With a dt parameter of 300 seconds, the diffusive_cnt code exhibits numerical oscillations at headwater locations. Though with a dt = 600 seconds, this behavior is minimized. Despite some numerical oscillations, the simulation does not blow up at any point. Numerical oscillations near headwaters are ultimately less concerning because the diffusive wave model is not intended for use in headwater areas. 
2. Ensure that DA is happening correctly at gage locations - should see same exact substitution result and similar nudge pattern as with Muskingum-Cunge
- passed
3. Results for all segments and reservoirs above gages should be identical to no DA simulation result
- passed

## TODO
- [x] copy capability to Tulane diffusive version, not just CNT
- [x] proof of concept figures/charts
- [x] comprehensive explaination
- [x] test that workflow still works when no DA parameters are specified in YAML (i.e. dw with reservoirs, no da)
- [x] clean cython code
- [x] use upstream boundary data at native time resolution - why sub-sample to timestep of qlaterals?
- [x] what is going on when nts is set to limit of available CHRTOUT files?
